### ISSUE TITLE
Update frequency/method and warm-up period

### ISSUE THREAD

**Issue by wuxmax** [Curt-Park/rainbow-is-all-you-need#52]
Hey!

First: Thanks a lot for the awesome tutorial! It is really great :)

I am currently building a Rainbow RL agent for the board game of Abalone based on your tutorial/implementation and [gym-abalone](https://github.com/towzeur/gym-abalone). You can have a look at it [here](https://github.com/wuxmax/rl4abalone).

While reading the [Rainbow paper](https://arxiv.org/abs/1710.02298), I found some discrepancies regarding when a learning step takes place and when the target network is updated:
| Issue | Paper | This repo |
| --- | ---| --- |
| 1 | training starts after warm-up period | training starts as soon as first batch is available |
| 2 | model update is performed every 4th agent step | update is performed at every agent step |
| 3 | target net is soft-updated at every training step? | target net is hard-updated every `target_update` steps |

With issue 2, I understand that in the paper each action selected by the agent is repeated four times and that every state is a concatenation of four frames. So from that perspective updating at every agent step might make sense.

With issue 3, I am not quite sure how it is done in the paper. It seems to be done like that in [this implementation](https://github.com/medipixel/rl_algorithms/blob/b3df31e62397e8ecef6973a5fd4696ee3d295b58/rl_algorithms/dqn/learner.py#L130), on which you build upon (right?).

Is there a reason for these differences?
I would be very happy to hear from you! :)

Best,
Max

**Comment by fmaschhur** [comment]
Working together with @wuxmax on this project, we were able to decide on these differences. These are our reasonings:

### Issue 1
Rainbow in fact does only start training after a certain warm-up period (80K frames), which is definitely needed when using Experience Replay.
We have implemented the same behaviour in our adaptation.

### Issue 2 
The stacking of 4 frames and repetition of actions for 4 frames in the original Rainbow paper hints that frequent updates may be preferred in settings comparable to Atari games. Since this tutorial does not implement stacking of frames there is no need to wait any amount of agent steps between updates. 
However, in the board-game setting this might be too frequent. Because of this we are implementing the amount of steps between each update as an hyperparameter which should probably be set to something between `batch_size/4` and `batch_size/2`.

### Issue 3
The original Rainbow paper suggests the implementation of hard-updates. Soft-updates are known to take more time to train, but offer higher stability in training.
Since Rainbow is already quite time-consuming to train, we decided to go with the original version.

Any response from the original authors of this repo is still highly appreciated!

**Comment by MrSyee** [comment]
Hi @wuxmax @fmaschhur !
I'm glad to hear that our repo is helpful to your project. These are our opinions for your question:

### Issue1
As you say, Training after warm-up period is an effective way for training Rainbow agent, especially in complex environments.
But Cart-Pole is a very simple environment. It is trained well without warm-up period.
We also use warm-up period to train Rainbow in other environments. (We can set warm-up period by `update_starts_from` in [other implement](https://github.com/medipixel/rl_algorithms/blob/3d106e5376031f60cc3a822f393612c9fb65467f/rl_algorithms/dqn/agent.py#L268))

### Issue2
Without frame stacking, model update period is also important. Training too frequently (like every step update) is bad because data in replay buffer is almost same after 1 step. So the interval between each update is helpful for training.
In this tutorial, for the same reason issue 1, Cart-pole is very simple and easy, so it shows good performance without update period.
(We can set update period by `train_freq` in [other implement](https://github.com/medipixel/rl_algorithms/blob/3d106e5376031f60cc3a822f393612c9fb65467f/rl_algorithms/dqn/agent.py#L269))

### Issue3
Soft target update is introduced in [DDPG paper](https://arxiv.org/pdf/1509.02971.pdf).
As you know, soft-update offer higher stability in training because the target is changed little by little.
We think soft target update is also effective for DQN empirically, so we use this way.

I hope your work goes well! Tell me the good news.
